Add ResizeAndComputePrimitives mutator to NewtonianEuler

nilsdeppe opened the pull request:
## Proposed changes

Exactly what the title says. This is used to compute the prims after we've decided we can go back to DG.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Add ResizeAndComputePrimitives mutator to NewtonianEuler

fmahebert commented on src/Evolution/Systems/NewtonianEuler/Subcell/ResizeAndComputePrimitives.hpp:
```diff
@@ -0,0 +1,73 @@
+ * `::Actions::MutateApply` action right after the
+ * `evolution::dg::subcell::Actions::TciAndSwitchToDg` action. In the case the
+ * active grid switches from subcell to DG, the resize and primitives are
+ * compute, while if the active grid remains subcell then the primitives are
```
"the resize and primitives are compute" -> "the primitives are resized and computed" ? Also clarify if the prims are computed pointwise from the DG cons vars, or by a reconstruction.

fmahebert commented on src/Evolution/Systems/NewtonianEuler/Subcell/ResizeAndComputePrimitives.hpp:
```diff
@@ -0,0 +1,73 @@
+ * `evolution::dg::subcell::Actions::TciAndSwitchToDg` action. In the case the
+ * active grid switches from subcell to DG, the resize and primitives are
+ * compute, while if the active grid remains subcell then the primitives are
+ * computed on the subcells.
```
Well this is in direct contradiction to the implementation, which does nothing in the "grid remains subcell" case and explains it with a nice comment. Can the nice comment also go here? :)

fmahebert commented on tests/Unit/Evolution/Systems/NewtonianEuler/Subcell/Test_ResizeAndComputePrimitives.cpp:
```diff
@@ -0,0 +1,110 @@
+      make_not_null(&box));
+
+  REQUIRE(db::get<::Tags::Variables<prim_tags>>(box).number_of_grid_points() ==
+          cons_vars.number_of_grid_points());
```
Why is this a `REQUIRE` vs a `CHECK`?

nilsdeppe replied:
because if this isn't true the check below is done on different grids and so is guaranteed to fail

fmahebert replied:
... but you want it to fail, so why is that bad? :P

nilsdeppe commented:
I've rebased and squashed in the rewording of the sentence and added a sentence that makes it clear that the primitives are recomputed from the reconstructed conserved variables. Thanks for the review :)

kidder commented on src/Evolution/Systems/NewtonianEuler/Subcell/ResizeAndComputePrimitives.hpp:
```diff
@@ -0,0 +1,75 @@
+#include "Evolution/DgSubcell/Tags/ActiveGrid.hpp"
+#include "Evolution/DgSubcell/Tags/Mesh.hpp"
+#include "Evolution/Systems/NewtonianEuler/Tags.hpp"
+#include "PointwiseFunctions/Hydro/EquationsOfState/EquationOfState.hpp"
```
forward declare?

kidder commented on src/Evolution/Systems/NewtonianEuler/Subcell/ResizeAndComputePrimitives.hpp:
```diff
@@ -0,0 +1,75 @@
+
+  template <size_t ThermodynamicDim>
+  static void apply(
+      gsl::not_null<Variables<
```
forward declare Variables

kidder commented on src/Evolution/Systems/NewtonianEuler/Subcell/ResizeAndComputePrimitives.cpp:
```diff
@@ -0,0 +1,76 @@
+                                                             : subcell_mesh)
+          .number_of_grid_points();
+  if (prim_vars->number_of_grid_points() != num_grid_points) {
+    ASSERT(active_grid == evolution::dg::subcell::ActiveGrid::Dg,
```
include for ASSERT

kidder commented on src/Evolution/Systems/NewtonianEuler/Subcell/ResizeAndComputePrimitives.cpp:
```diff
@@ -0,0 +1,76 @@
+
+#include "Evolution/Systems/NewtonianEuler/Subcell/ResizeAndComputePrimitives.hpp"
+
+#include <algorithm>
```
unused

kidder reviewed (changes requested):
you can squash

nilsdeppe commented:
Okay, I've rebased and squashed. Thanks for the reviews :)
